editoast/fga_migrations/src/lib.rs

```diff
@@ -0,0 +1,345 @@
+
+#[derive(fga::User, Debug, Clone, PartialEq, Eq, Hash)]
+/// Effectively a constant. We don't need anything from it but Openfga cannot store anything else
+/// than tuples so we need to define another type and a relation in addition to the Migration type.
+struct Editoast;
+
+fga::relations! {
+    Migration {
+        apply: Editoast
+    }
+}
+
+impl Type for Editoast {
+    const NAMESPACE: &'static str = "editoast";
+
+    fn id(&self) -> impl ToString {
+        "editoast"
+    }
+}
+
+impl FromStr for Editoast {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "editoast" {
+            return Ok(Editoast);
+        }
+        Err(format!("not a valid service: `{}`", s))
+    }
+}
+
+impl Type for Migration {
+    const NAMESPACE: &'static str = "migration";
+
+    fn id(&self) -> impl ToString {
+        format!("{}-{}", self.0, self.1)
+    }
+}
+
+impl FromStr for Migration {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (migration_number, model_hash) = s
+            .split_once("-")
+            .ok_or("invalid migration identifier".to_string())?;
+        let migration_number = migration_number
+            .parse::<usize>()
+            .map_err(|_| "invalid migration number".to_string())?;
+        Ok(Migration(migration_number, model_hash.to_string()))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum MigrationError {
+    #[error("Invalid migration name: `{0}`")]
+    InvalidMigrationName(String),
+    #[error(transparent)]
+    InitializationError(#[from] InitializationError),
+    #[error(transparent)]
+    RequestFailure(#[from] RequestFailure),
+    #[error("Inconsistent migrations store state")]
+    InconsistantMigrationStore,
+}
+
+pub fn get_migration_index(target_migration: &TargetMigration) -> Result<usize, MigrationError> {
+    match target_migration {
+        TargetMigration::Latest => match OPENFGA_MIGRATIONS.is_empty() {
+            true => unreachable!("the migration list is static and should never be empty"),
+            false => Ok(OPENFGA_MIGRATIONS.len() - 1),
+        },
+        TargetMigration::Name(migration_name) => {
+            match OPENFGA_MIGRATIONS.get_index_of(&migration_name.as_str()) {
+                Some(migration_index) => Ok(migration_index),
+                None => Err(MigrationError::InvalidMigrationName(
+                    migration_name.to_string(),
+                ))?,
+            }
+        }
+    }
+}
+
+#[instrument(name = "migration_run")]
+pub async fn run_migrations(
+    mut client_authz: Client,
+    mut client_migrations: Client,
+    target_migration: TargetMigration,
+) -> Result<(), MigrationError> {
+    // Update the migration store model
+    let migration_number = get_migration_index(&target_migration)?;
+    let migration_model_dsl = OPENFGA_MIGRATIONS[migration_number];
+    info!("Retrieved migration DSL model: {}", migration_model_dsl);
```

Comment:
```suggestion
    info!("Retrieved migration DSL model: {migration_name}[{migration_id}]");
```

printing the whole model floods the console

Reply:
tl;dr: I prefer using structured logging here if that's fine with you

I added the migration name and number as key-value pairs in the tracing event instead of formatting them in the message as suggested. It gives less false assumptions to someone reading the logs about how they are represented in the data structures imo. `migration_name[migration_id]` suggests that we have a map that associates a migration name (key) with its identifier (value) when in reality we can use either as respectively a key or index to find the related migration model (value) in the indexed map that stores the DSL models.

Reply:
+1 about structured logging, but unless I'm wrong, this is still a formatted string:

```rust
info!(
    "Retrieved DSL model of migration {}: `{}`",
    migration_number, migration_name
);

// equivalent to
info!(
    "Retrieved DSL model of migration {migration_number}: `{migration_name}`",
);
```

Key-value pairs must be placed *before* the string, like so:

```rust
info!(
	migration_number,
	migration_name
      "Retrieved DSL model",
  );
```

I let you change that if that's correct and you agree, and resolve this comment.

Reply:
I think you viewed an outdated commit. That was already the case in the latest commit with that peculiar cargo formatting:
```rs
    info!(
        migration_number,
        migration_name, "Retrieved DSL model of target migration",
    );

```
Resolving the thread